Java frontend: create String and Class literals earlier

smowton opened the pull request:
This creates String and Class literals ahead of running java_bytecode_convert_method
instead of creating them in an after-the-fact fixup in typecheck_expr as previously.
These now exist before method conversion begins and __CPROVER_initialize is created,
which means the String literals are correctly initialised even when using incremental
method conversion.

This can be a little wasteful if they are only referred to from methods that would never
truly be invoked, but preliminary experiments suggest that creating ~10000 Strings takes
under a second, and that Java lazy loading can quickly delete ones that are not in fact
used.

This "right first time" approach to Java conversion has similar benefits to https://github.com/diffblue/cbmc/pull/1764, and makes incremental function loading easier.

thk123 commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -195,6 +196,87 @@ bool java_bytecode_languaget::parse(
+            new_class_symbol.name = symbol_expr.get_identifier();
+            new_class_symbol.type = symbol_expr.type();
+            // name without "java::" prefix:
+            new_class_symbol.base_name =
```
```C++
INVARIANT(has_suffix(new_class_symbol.name, "java::")
```

thk123 commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -195,6 +196,87 @@ bool java_bytecode_languaget::parse(
+
+        if(ldc_arg0.id() == ID_type)
+        {
+          irep_idt class_id = ldc_arg0.type().get(ID_identifier);
```
I'd be tempted to pull this into a `get_or_create_class_constant` or something

thk123 commented on src/java_bytecode/java_string_literals.cpp:
```diff
@@ -0,0 +1,197 @@
+#include <iomanip>
+#include <sstream>
+
+static std::string escape_non_alnum(const std::string &toescape)
```
I appreciate you are just moving a function, but now would be a great time to add some doxygen to this function (I for one have no idea what an alnum is and why I'd want to escape it). 
Nit: `toescape` -> `to_escape`

smowton commented:
TG bump: https://github.com/diffblue/test-gen/pull/1413

thk123 reviewed (dismissed):
TG pointer bump failing on a bunch of string tests...

smowton commented:
@thk123 changes applied, and I think I've found the cause of the test failures.

@romainbrenguier @allredj it appears when String constant globals are made earlier and string-refinement is in play, then two ancillary globals are created:

1. `Literal.xyz.constarray = { 'x', 'y', 'z' }`. This was being eliminated by ci-lazy-methods' unused-globals sweep, which I fix here by augmenting it slightly to notice that Literal.xyz's initialiser referring to it implies it is needed.

2. `Literal.xyz.return_value = associate_array_to_pointer({'x', 'y', 'z'}, &Literal.xyz)`. This was stranger -- if I'm reading it correctly, this is never used and is just there to force that associate function to "run" (except it is declared as a mathematical function, so it usually isn't supposed to have side-effects). I've worked around this by initialising Literal.xyz with `Literal.xyz.return_value, &Literal.xyz.constarray` (as in a C comma-expression), intending to indicate that "if you need Literal.xyz then you also need Literal.xyz_return_value", which seems to be enough to convince everybody that the return_value expression is needed. 

However this is pretty fragile, since the mathematical function construct is being somewhat abused here since its return value is unimportant, and I would expect a future slightly-cleverer unused-globals analysis to delete it again.

In summary I recommend coming up with a better way to express the array <-> String pointer association that doesn't use a mathematical-function-with-side-effect.

smowton commented:
Fixed failure, was caused by this patch promoting `@class_model` objects to constant globals (which they always should have been). They contain HashTable members which get a `cproverNondetInitialize` call, which in turn could fail because the object factory could run the init function at the wrong time. I'll file a separate PR for the fix; this PR for the time being depends on that one.

smowton commented:
The nondetInitialize fix has been split into a separate PR: https://github.com/diffblue/cbmc/pull/1769

LAJW commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
+    symbolt new_class_symbol;
+    new_class_symbol.name = symbol_expr.get_identifier();
+    new_class_symbol.type = symbol_expr.type();
+    // name without "java::" prefix:
```
I think this comment should be removed.

smowton replied:
Not done; without it it is not clear what we're trying to clip by removing 6 chars.

LAJW commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
+      if(instruction.statement == "ldc" ||
+         instruction.statement == "ldc2" ||
+         instruction.statement == "ldc_w" ||
+         instruction.statement == "ldc2_w")
```
It'd be great if you could pull this condition into a function. I'm not sure what any of those magic values mean.

smowton replied:
Added a comment. No other Java bytecode opcodes have such functions, so I'll leave these ones plain as well for now.

LAJW commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
+        // onto the stack.
+
+        // 3) Pushing a Class constant causes a reference to a java.lang.Class
+        // to be pushed onto the operand stack
```
These comments describe "what" and not why, which renders them unnecessary.

smowton replied:
Not done; this describes the semantics of the opcode more clearly than reading the code.

LAJW commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
+        // 3) Pushing a Class constant causes a reference to a java.lang.Class
+        // to be pushed onto the operand stack
+
+        exprt &ldc_arg0 = instruction.args[0];
```
I'd mark it `const exprt &ldc_arg0` here, and then assign `instruction.args[0]` to a closure. Would make it clear that this is not modified in called functions, only reassigned.

smowton replied:
Split it into a separate function.

LAJW commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -317,6 +418,23 @@ bool java_bytecode_languaget::typecheck(
+  // This means that when running with lazy loading, we will be aware of these
+  // literal globals' existence when __CPROVER_initialize is generated in
+  // `generate_support_functions`.
+  std::size_t before_constant_globals_size = symbol_table.symbols.size();
```
`const`, please?

LAJW commented on src/java_bytecode/java_string_literals.cpp:
```diff
@@ -0,0 +1,211 @@
+              << (unsigned int)ch;
+  }
+  return escaped.str();
+}
```
Moving those functions into a separate file should be done in a different commit. Right now it looks similar, but not identical.

Edit: Commit separated, I commend.

LAJW commented on src/java_bytecode/java_string_literals.cpp:
```diff
@@ -0,0 +1,211 @@
+    "in the symbol table, so adding it should succeed");
+
+  return new_symbol.symbol_expr();
+}
```
This function is way too long (but I suppose it would be acceptable to put it here as a move-only commit).

Edit: Fine, as it has been put into a move-only commit.

LAJW reviewed (changes requested):
The only thing that kills it for me is the fact, that moving code around isn't done in a dedicated commit.

smowton committed:
Move string literal initialisation into separate file

smowton committed:
Java frontend: create String an Class literals earlier

This creates String and Class literals ahead of running java_bytecode_convert_method
instead of creating them in an after-the-fact fixup in typecheck_expr as previously.
These now exist before method conversion begins and __CPROVER_initialize is created,
which means the String literals are correctly initialised even when using incremental
method conversion.

This can be a little wasteful if they are only referred to from methods that would never
truly be invoked, but preliminary experiments suggest that creating ~10000 Strings takes
under a second, and that Java lazy loading can quickly delete ones that are not in fact
used.

Java CI lazy methods is augmented a little to notice that if we need global A, and global
A is initialised with a reference to global B, then we need global B as well. We appear to
have got away wtih this before because the inter-global references were introduced after
CI lazy methods finished loading methods.

smowton commented:
@LAJW split the move apart from the substantial commit, and implemented those suggestions not otherwise responded to above.

smowton committed:
CI lazy methods: scan global initialisers for global references

CI lazy methods now notices cases like global g1 having initialiser &g2, in which case
having a reference to global g1 implies a reference to global g2.

Currently this only happens in the string solver in Java, where a literal's initialiser
references a constant array global.

smowton commented:
@LAJW please re-review; @peterschrammel @thk123 @tautschnig this is ready for review

thk123 reviewed (approved):
Change looks good, TG pointer bump is passing

LAJW commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -256,6 +257,118 @@ static void infer_opaque_type_fields(
+        get_or_create_string_literal_symbol(
+          ldc_arg0, symbol_table, string_refinement_enabled));
+  }
+  else
```
This else is redundant.

LAJW reviewed (approved):
It's fine.
